gcs/constraints/circuit/circuit.hh

```diff
@@ -0,0 +1,130 @@
+
+
+#ifndef GLASGOW_CONSTRAINT_SOLVER_GUARD_GCS_CIRCUIT_HH
+#define GLASGOW_CONSTRAINT_SOLVER_GUARD_GCS_CIRCUIT_HH
+
+#include <gcs/constraint.hh>
+#include <gcs/innards/proof-fwd.hh>
+#include <gcs/innards/proof.hh>
+
+#include <gcs/innards/state.hh>
+#include <gcs/variable_id.hh>
+#include <list>
+#include <map>
+#include <optional>
+#include <vector>
+
+namespace gcs
+{
+    struct ProofFlagData
+    {
+        std::string comment_name;
+        innards::ProofFlag flag;
+        innards::ProofLine forwards_reif_line;
+        innards::ProofLine backwards_reif_line;
+        std::map<long, innards::ProofLine> neq_lines;
+    };
+
+    struct PosVarLineData
+    {
+        innards::ProofLine leq_line;
+        innards::ProofLine geq_line;
+    };
+
+    struct PosVarData
+    {
+        std::string comment_name;
+        innards::ProofOnlySimpleIntegerVariableID var;
+        std::map<long, PosVarLineData> plus_one_lines;
+    };
+
+    struct PosAllDiffData
+    {
+        std::map<long, innards::ProofLine> at_most_1_lines;
+        std::map<long, innards::ProofLine> at_least_1_lines;
+    };
+
+    using ProofFlagDataMap = std::map<long, std::map<long, ProofFlagData>>;
+    using PosVarDataMap = std::map<long, PosVarData>;
+
+    struct ShiftedPosDataMaps
+    {
+        std::map<long, ProofFlagData> greater_than;
+        ProofFlagDataMap shifted_pos_eq;
+        ProofFlagDataMap shifted_pos_geq;
+    };
+
+    struct SCCOptions
+    {
+        bool prune_root = true;
+        bool prune_skip = true;
+        bool fix_req = true;
+        bool prune_within = true;
+        bool prove_using_dominance = false;
+        bool enable_comments = true;
+    };
+
+    /**
+     * \brief Circuit constraint: requires the variables, representing graph nodes, take values
+     * such that each variable's value represents the index of the next node in a tour that visits
+     * all variables.
+     *
+     * \ingroup Constraints
+     */
+    class CircuitBase : public Constraint
```

Comment:
Maybe more header files would be useful? So `CircuitBase`, `CircuitPrevent`, etc get their own headers, and `circuit.hh` just includes them and has the final `using` declaration? I find it less confusing if there's a direct correspondence between the `cc` and `hh` files. Also some of our source files are probably getting a bit long and this might help break them up.

Reply:
Clarifying, I mean, `constraints/circuit.hh` rather than `constraints/circuit/circuit.hh`